assets/prometheus-k8s/rules.yaml

```diff
@@ -380,6 +380,14 @@ spec:
       record: node_role_os_version_machine:cpu_capacity_sockets:sum
     - expr: clamp_max(sum(alertmanager_notifications_total),1)
       record: cluster:alertmanager_routing_enabled:max
+    - alert: KubePodOOMWarning
+      annotations:
+        message: Pod {{ $labels.namespace  }}/{{ $labels.pod  }} ({{ $labels.container  }})
+          is experiencing out of memory termination.
+      expr: sum(max(kube_pod_container_status_last_terminated_reason{reason="OOMKilled",namespace=~"(openshift-.*|kube-.*|default|logging)",job="kube-state-metrics"}))
```

Comment:
kube_pod_container_status_last_terminated_reason only contains as it says last terminated reason, as its a gauge. This would alert everytime we have a OOMKilled which seems very drastic and would lead to alert fatigue. I would prefer to collect the OOM reasons in a counter, but since we do not have this metric right now I would suggest something along the lines of this, note this query is just a rough idea and has not been tested:

```
kube_pod_container_status_last_terminated_reason{reason=“OOMKilled”,} == 1 
    and on(container) rate(kube_pod_container_status_restarts_total[5m]) * 300 > 1
```

I think while this does not mean all of those containers were terminated because OOM, it still gives us a better indication that something is actually wrong in the cluster.

Second thing, we tend to like to contribute all alerts in upstream, so once we decide on this, we should have this be in the upstream repo and we can bring it in. https://github.com/kubernetes-monitoring/kubernetes-mixin

cc @openshift/openshift-team-monitoring for thoughts and improvements :)

Reply:
This has also been discussed numerous times on sig-instrumentation. The consensus was that ideally the kubelet or container runtime could expose a counter of oomkilled containers, that way we would actually have a reliable counter instead of this heuristic. That said, the heuristic suggested by @LiliC might be "good enough :tm: ".